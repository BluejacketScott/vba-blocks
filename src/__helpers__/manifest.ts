import { dir as FIXTURES } from '../../tests/__fixtures__';
import { Manifest } from '../manifest';
import { isPathDependency } from '../manifest/dependency';
import { relative } from '../utils/path';

export function normalizeManifest(manifest: Manifest, relativeTo: string = FIXTURES): Manifest {
  for (const src of manifest.src) {
    src.path = normalizePath(src.path, relativeTo);

    if (src.binary) {
      src.binary = normalizePath(src.binary, relativeTo);
    }
  }
  for (const dependency of manifest.dependencies) {
    if (isPathDependency(dependency)) {
      dependency.path = normalizePath(dependency.path, relativeTo);
    }
  }
  for (const src of manifest.devSrc) {
    src.path = normalizePath(src.path, relativeTo);

    if (src.binary) {
      src.binary = normalizePath(src.binary, relativeTo);
    }
  }
  for (const dependency of manifest.devDependencies) {
    if (isPathDependency(dependency)) {
      dependency.path = normalizePath(dependency.path, relativeTo);
    }
  }

  if (manifest.target) {
    manifest.target.path = normalizePath(manifest.target.path, relativeTo);
  }

  return manifest;
}

function normalizePath(path: string, relativeTo: string = FIXTURES): string {
  return relative(relativeTo, path);
}
