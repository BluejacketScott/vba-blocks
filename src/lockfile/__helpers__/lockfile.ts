export function normalizeLockfile(lockfile: string): string {
	return lockfile.replace(/vba\-blocks v\d.\d.\d/g, "vba-blocks v#.#.#");
}
