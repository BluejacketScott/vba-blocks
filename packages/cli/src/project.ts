import { join } from 'path';
import env from './env';
import { Config, loadConfig } from './config';
import { Manifest, Source, Reference, loadManifest } from './manifest';
import { Workspace } from './workspace';
import SourceManager, { Registration } from './sources';
import resolve, { DependencyGraph, getRegistration } from './resolve';

import { parallel } from './utils';

export interface Project {
  manifest: Manifest;
  workspace: Workspace;
  packages: DependencyGraph;

  config: Config;
  paths: {
    root: string;
    dir: string;
    build: string;
    backup: string;
  };
}

export async function loadProject(dir?: string): Promise<Project> {
  const manifest = await loadManifest(dir || env.cwd);

  // TODO Load workspace
  const workspace: Workspace = {
    root: manifest,
    members: []
  };

  const config = await loadConfig(manifest, workspace);
  const packages = await resolve(config, workspace);

  const paths = {
    root: workspace.root.dir,
    dir: manifest.dir,
    build: join(manifest.dir, 'build'),
    backup: join(manifest.dir, 'build', '.backup')
  };

  return {
    manifest,
    workspace,
    packages,
    config,
    paths
  };
}

export async function fetchDependencies(project: Project): Promise<Manifest[]> {
  const manager = new SourceManager(project.config);
  const manifests = await parallel(
    project.manifest.dependencies,
    async dependency => {
      const registration = getRegistration(project.packages, dependency);
      const path = await manager.fetch(registration);
      const manifest = await loadManifest(path);

      return manifest;
    }
  );

  return manifests;
}
