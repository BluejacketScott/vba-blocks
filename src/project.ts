import env from './env';
import { join, normalize } from './utils/path';
import { tmpFolder } from './utils/fs';
import parallel from './utils/parallel';
import { Config, loadConfig } from './config';
import { Manifest, loadManifest } from './manifest';
import { Workspace, loadWorkspace } from './workspace';
import { fetch } from './sources';
import resolve, { DependencyGraph } from './resolve';
import { readLockfile, isLockfileValid } from './lockfile';

const debug = require('debug')('vba-blocks:project');

export interface Project {
  manifest: Manifest;
  workspace: Workspace;
  packages: DependencyGraph;

  config: Config;
  paths: {
    root: string;
    dir: string;
    build: string;
    backup: string;
    staging: string;
  };
  has_dirty_lockfile: boolean;
}

/**
 * Load project from given directory / cwd
 *
 * 1. Load manifest in dir
 * 2. Load workspace from manifest
 * 3. Load lockfile
 * 4. Resolve dependencies
 */
export async function loadProject(dir: string = env.cwd): Promise<Project> {
  const manifest = await loadManifest(dir);

  const config = await loadConfig();
  const workspace = await loadWorkspace(manifest);
  const lockfile = await readLockfile(workspace.root.dir);

  // Resolve packages from lockfile or from sources
  const has_dirty_lockfile = !lockfile || !isLockfileValid(lockfile, workspace);
  debug(
    !has_dirty_lockfile
      ? 'Loading packages from lockfile'
      : 'Resolving packages'
  );

  const packages = !has_dirty_lockfile
    ? lockfile!.packages
    : await resolve(config, workspace, lockfile ? lockfile.packages : []);

  const paths = {
    root: workspace.root.dir,
    dir: manifest.dir,
    build: join(manifest.dir, 'build'),
    backup: join(manifest.dir, 'build', '.backup'),
    staging: await tmpFolder({ dir: env.staging })
  };

  return {
    manifest,
    workspace,
    packages,
    config,
    paths,
    has_dirty_lockfile
  };
}

export interface FetchProject {
  manifest: Manifest;
  packages: DependencyGraph;
  config: Config;
}

/**
 * Fetch all dependencies for project
 * (based on already resolved project.packages)
 *
 * After sources complete fetches, manifests are loaded and returned for each package
 */
export async function fetchDependencies(
  project: FetchProject
): Promise<Manifest[]> {
  const manifests = await parallel(
    project.packages,
    async registration => {
      const path = await fetch(project.config.sources, registration);
      const manifest = await loadManifest(path);

      return manifest;
    },
    { progress: env.reporter.progress('Fetch dependencies') }
  );

  return manifests;
}

export interface ProjectOptions {
  type?: 'project' | 'package';
}

/**
 * Initialize new project
 *
 * Minimum requirements: name and dir
 */
export async function initProject(
  name: string,
  dir: string,
  options: ProjectOptions = {}
): Promise<Project> {
  dir = normalize(dir);
  const { type = 'project' } = options;

  const config = await loadConfig();

  // TODO load defaults from config
  const version = '0.1.0';
  const authors: string[] = [];
  const license = 'MIT';

  // Manually generate manifest and project
  // (may be included in project in the future)
  const manifest: Manifest = {
    type,
    name,
    version,
    metadata: {
      authors,
      publish: type === 'package',
      license
    },
    dependencies: [],
    src: [],
    references: [],
    targets: [],
    dir
  };

  const workspace = await loadWorkspace(manifest);

  const project: Project = {
    manifest,
    workspace,
    packages: [],
    config,
    paths: {
      root: workspace.root.dir,
      dir,
      build: join(dir, 'build'),
      backup: join(dir, 'build', '.backup'),
      staging: await tmpFolder({ dir: env.staging })
    },
    has_dirty_lockfile: true
  };

  return project;
}
