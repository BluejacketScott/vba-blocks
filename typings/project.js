import env from './env';
import { join, normalize } from './utils/path';
import { tmpFolder } from './utils/fs';
import parallel from './utils/parallel';
import { loadConfig } from './config';
import { loadManifest } from './manifest';
import { loadWorkspace } from './workspace';
import { fetch } from './sources';
import resolve from './resolve';
import { readLockfile, isLockfileValid } from './lockfile';
import { fetchingDependencies } from './messages';
const debug = require('debug')('vba-blocks:project');
/**
 * Load project from given directory / cwd
 *
 * 1. Load manifest in dir
 * 2. Load workspace from manifest
 * 3. Load lockfile
 * 4. Resolve dependencies
 */
export async function loadProject(dir = env.cwd) {
    const manifest = await loadManifest(dir);
    const config = await loadConfig();
    const workspace = await loadWorkspace(manifest);
    const lockfile = await readLockfile(workspace.root.dir);
    // Resolve packages from lockfile or from sources
    const has_dirty_lockfile = !lockfile || !(await isLockfileValid(lockfile, workspace));
    debug(!has_dirty_lockfile ? 'Loading packages from lockfile' : 'Resolving packages');
    const packages = !has_dirty_lockfile
        ? lockfile.packages
        : await resolve(config, workspace, lockfile ? lockfile.packages : []);
    const paths = {
        root: workspace.root.dir,
        dir: manifest.dir,
        build: join(manifest.dir, 'build'),
        backup: join(manifest.dir, 'build', '.backup'),
        staging: await tmpFolder({ dir: env.staging })
    };
    return {
        manifest,
        workspace,
        packages,
        config,
        paths,
        has_dirty_lockfile
    };
}
/**
 * Fetch all dependencies for project
 * (based on already resolved project.packages)
 *
 * After sources complete fetches, manifests are loaded and returned for each package
 */
export async function fetchDependencies(project) {
    const manifests = await parallel(project.packages, async (registration) => {
        const path = await fetch(project.config.sources, registration);
        const manifest = await loadManifest(path);
        return manifest;
    }, { progress: env.reporter.progress(fetchingDependencies()) });
    return manifests;
}
/**
 * Initialize new project
 *
 * Minimum requirements: name and dir
 */
export async function initProject(name, dir, options = {}) {
    dir = normalize(dir);
    const { type = 'project' } = options;
    const config = await loadConfig();
    // TODO load defaults from config
    const version = '0.0.0';
    const authors = [];
    const license = 'UNLICENSED';
    const __temp_defaults = type === 'package' ? [] : ['publish'];
    // Manually generate manifest and project
    // (may be included in project in the future)
    const manifest = {
        type,
        name,
        version,
        metadata: {
            authors,
            publish: type === 'package',
            license,
            __temp_defaults
        },
        dependencies: [],
        src: [],
        references: [],
        dir
    };
    const workspace = await loadWorkspace(manifest);
    const project = {
        manifest,
        workspace,
        packages: [],
        config,
        paths: {
            root: workspace.root.dir,
            dir,
            build: join(dir, 'build'),
            backup: join(dir, 'build', '.backup'),
            staging: await tmpFolder({ dir: env.staging })
        },
        has_dirty_lockfile: true
    };
    return project;
}
