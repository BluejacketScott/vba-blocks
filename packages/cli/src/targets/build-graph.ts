import { Source, Reference } from '../manifest';
import { Project, fetchDependencies, loadManifests } from '../project';
import { BuildOptions } from './build-target';

export interface BuildGraph {
  src: Source[];
  references: Reference[];
}

export async function createBuildGraph(
  project: Project,
  options: BuildOptions
): Promise<BuildGraph> {
  const src: Map<string, Source> = new Map();
  const references: Map<string, Reference> = new Map();

  const manifests = project.manifests || (await loadManifests(project));

  for (const manifest of manifests) {
    for (const source of manifest.src) {
      const { name } = source;
      if (src.has(name))
        throw new Error(`Conflicting source named "${source.name}"`);
      src.set(name, source);
    }
    for (const reference of manifest.references) {
      const { name, guid } = reference;
      if (references.has(name) && references.get(name)!.guid !== guid)
        throw new Error(`Conficting reference named "${reference.name}"`);
      references.set(name, reference);
    }
  }

  return { src: [...src.values()], references: [...references.values()] };
}
