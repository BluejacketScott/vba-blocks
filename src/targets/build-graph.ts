import { Manifest, Source, Reference } from '../manifest';
import { Project } from '../project';
import { BuildOptions } from './build-target';

export interface BuildGraph {
  src: Source[];
  references: Reference[];
}

export async function createBuildGraph(
  project: Project,
  dependencies: Manifest[],
  options: BuildOptions
): Promise<BuildGraph> {
  const src: Map<string, Source> = new Map();
  const references: Map<string, Reference> = new Map();
  const manifests = [project.manifest, ...dependencies];

  for (const manifest of manifests) {
    for (const source of manifest.src) {
      const { name } = source;
      if (src.has(name))
        throw new Error(`Conflicting source named "${source.name}"`);
      src.set(name, source);
    }
    for (const reference of manifest.references) {
      const { name, guid } = reference;
      if (references.has(name) && references.get(name)!.guid !== guid)
        throw new Error(`Conficting reference named "${reference.name}"`);
      references.set(name, reference);
    }
  }

  return { src: [...src.values()], references: [...references.values()] };
}
