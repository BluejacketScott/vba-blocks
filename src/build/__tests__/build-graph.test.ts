import { setup, reset } from '../../../tests/__helpers__/project';
import { standard } from '../../../tests/__fixtures__';
import { loadBuildGraph, BuildGraph } from '../build-graph';
import { Component } from '../component';

afterEach(reset);

test('should create build graph', async () => {
  const { project, dependencies } = await setup(standard);

  const graph = await loadBuildGraph(project, dependencies);
  expect(normalizeGraph(graph)).toMatchSnapshot();
});

function normalizeGraph(graph: BuildGraph) {
  const { name, components, references } = graph;

  return {
    name,
    components: components.map(toComponent),
    references
  };
}

export function truncate(value: string): string {
  return value.length > 200 ? `${value.slice(0, 200)}...` : value;
}

export function toComponent(component: Component) {
  return {
    name: component.name,
    type: component.type,
    filename: component.filename,
    binary_path: component.binary_path,
    code: truncate(component.code)
  };
}
