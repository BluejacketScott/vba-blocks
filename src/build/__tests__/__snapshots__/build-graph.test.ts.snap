// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should create build graph 1`] = `
Object {
  "components": Array [
    Object {
      "binary_path": undefined,
      "code": "VERSION 1.0 CLASS
BEGIN
  MultiUse = -1  'True
END
Attribute VB_Name = \\"ThisWorkbook\\"
Attribute VB_GlobalNameSpace = False
Attribute VB_Creatable = False
Attribute VB_PredeclaredId = True
Attr...",
      "filename": "ThisWorkbook.cls",
      "name": "ThisWorkbook",
      "type": "class",
    },
    Object {
      "binary_path": undefined,
      "code": "VERSION 1.0 CLASS
BEGIN
  MultiUse = -1  'True
END
Attribute VB_Name = \\"Sheet1\\"
Attribute VB_GlobalNameSpace = False
Attribute VB_Creatable = False
Attribute VB_PredeclaredId = True
Attribute ...",
      "filename": "Sheet1.cls",
      "name": "Sheet1",
      "type": "class",
    },
    Object {
      "binary_path": undefined,
      "code": "VERSION 1.0 CLASS
BEGIN
  MultiUse = -1  'True
END
Attribute VB_Name = \\"Sheet2\\"
Attribute VB_GlobalNameSpace = False
Attribute VB_Creatable = False
Attribute VB_PredeclaredId = True
Attribute ...",
      "filename": "Sheet2.cls",
      "name": "Sheet2",
      "type": "class",
    },
    Object {
      "binary_path": undefined,
      "code": "VERSION 1.0 CLASS
BEGIN
  MultiUse = -1  'True
END
Attribute VB_Name = \\"Sheet3\\"
Attribute VB_GlobalNameSpace = False
Attribute VB_Creatable = False
Attribute VB_PredeclaredId = True
Attribute ...",
      "filename": "Sheet3.cls",
      "name": "Sheet3",
      "type": "class",
    },
    Object {
      "binary_path": "C:/dev/vba-blocks/tests/__fixtures__/projects/standard/src/UserForm1.frx",
      "code": "VERSION 5.00
Begin {C62A69F0-16DC-11CE-9E98-00AA00574A4F} UserForm1 
   Caption         =   \\"UserForm1\\"
   ClientHeight    =   3040
   ClientLeft      =   110
   ClientTop       =   450
   Clien...",
      "filename": "UserForm1.frm",
      "name": "UserForm1",
      "type": "form",
    },
    Object {
      "binary_path": undefined,
      "code": "Attribute VB_Name = \\"Validation\\"
Public Function Validate(Args As Variant) As String
    On Error GoTo ErrorHandling

    Dim Errors As New Collection

    If ThisWorkbook.VBProject.VBComponents...",
      "filename": "Validation.bas",
      "name": "Validation",
      "type": "module",
    },
    Object {
      "binary_path": undefined,
      "code": "VERSION 1.0 CLASS
BEGIN
  MultiUse = -1  'True
END
Attribute VB_Name = \\"Class1\\"
Attribute VB_GlobalNameSpace = False
Attribute VB_Creatable = False
Attribute VB_PredeclaredId = False
Attribute...",
      "filename": "Class1.cls",
      "name": "Class1",
      "type": "class",
    },
    Object {
      "binary_path": undefined,
      "code": "Attribute VB_Name = \\"JsonConverter\\"
''
' VBA-JSON v2.3.0
' (c) Tim Hall - https://github.com/VBA-tools/VBA-JSON
'
' JSON Converter for VBA
'
' Errors:
' 10001 - JSON parse error
'
' @class J...",
      "filename": "JsonConverter.bas",
      "name": "JsonConverter",
      "type": "module",
    },
    Object {
      "binary_path": undefined,
      "code": "VERSION 1.0 CLASS
BEGIN
  MultiUse = -1  'True
END
Attribute VB_Name = \\"Dictionary\\"
Attribute VB_GlobalNameSpace = False
Attribute VB_Creatable = False
Attribute VB_PredeclaredId = False
Attri...",
      "filename": "Dictionary.cls",
      "name": "Dictionary",
      "type": "class",
    },
  ],
  "name": "VBAProject",
  "references": Array [],
}
`;
